app/models/compute_resources/foreman/model/vmware.rb

```diff
@@ -499,15 +499,57 @@ def clone_vm(raw_args)
     end
 
     def console(uuid)
+      return vnc_console(uuid) if console_type == 'vnc'
+      webmks_console(uuid)
+    end
+
+    def vnc_console(uuid)
       vm = find_vm_by_uuid(uuid)
       raise "VM is not running!" unless vm.ready?
-      #TOOD port, password
-      #NOTE this requires the following port to be open on your ESXi FW
       values = { :port => unused_vnc_port(vm.hypervisor), :password => random_password, :enabled => true }
       vm.config_vnc(values)
       WsProxy.start(:host => vm.hypervisor, :host_port => values[:port], :password => values[:password]).merge(:type => 'vnc')
     end
 
+    def webmks_console(uuid)
+      vm = get_raw_vm(uuid)
+      ticket = webmks_ticket_for_raw_vm(vm)
+      host_url = URI::Generic.build(:scheme => 'https',
+                                    :host   => ticket.host,
+                                    :port   => ticket.port).to_s
+      console_url = URI::Generic.build(:scheme => 'wss',
+                                       :host   => ticket.host,
+                                       :port   => ticket.port,
+                                       :path   => "/ticket/#{ticket.ticket}").to_s
+      {
+        :name => vm.summary.config.name,
+        :host => ticket.host,
+        :extend_content_security_policy => ["wss://#{ticket.host}"],
+        :host_url => host_url,
+        :console_url => console_url,
+        :type => 'webmks'
+      }
+    end
+
+    # requests a webmks ticket for a vm
+    # uses pure rbvmomi to speed up ticket aquisition
```

Comment:
TBH I'm not super excited about this one, this has bitten us in the past multiple times.

I understand its easier to write the code directly in foreman vs add another layer in fog, but it later on it comes back as things you didnt think of (e.g. @orrabin was unable to cache ovirt stuff because of similar "feature" as it ended up pulling ssl layers that can't be serialized etc)